jstz_node/src/services/logs/broadcaster.rs

```diff
@@ -50,42 +51,57 @@ impl Broadcaster {
     async fn remove_stale_clients(&self) {
         let clients = self.clients.lock().clone();
 
-        let mut responsive_clients = Vec::new();
-
-        for client in clients {
-            if client
-                .send(sse::Event::Comment("ping".into()))
-                .await
-                .is_ok()
-            {
-                responsive_clients.push(client);
+        let mut responsive_clients: HashMap<Address, Vec<Sender<Event>>> = HashMap::new();
+
+        for (address, senders) in clients {
+            for sender in senders {
+                if sender
+                    .send(sse::Event::Comment("ping".into()))
+                    .await
+                    .is_ok()
+                {
+                    // TODO: why cannot we avoid cloning? clients are already moved?
+                    responsive_clients
+                        .entry(address.clone())
+                        .or_insert(Vec::default())
+                        .push(sender);
+                }
```

Comment:
To understand the borrowing issue here, we can annotate the owners as follows:
```rust
let clients = self.clients.lock().clone(); // <-- owned HashMap<Address, Vec<..>>

// ...

for (address, senders) in clients { // client moved here into iterator, address & sender are owned
  for sender in senders { // senders moved here, sender is owned
    if sender.send(..).await.is_ok() { // sender is immutably borrowed here
      responsive_clients
        .entry(address) // address is moved here, BUT it can only be moved once and Rust cannot ensure this loop only iterates once!
        .or_default()
        .push(sender);
    }
  }
}
```

Thus we get a borrow checking error. To satisfy the borrow checker, we need to hoist the entry call:
```rust

        for (address, senders) in clients {
            let responsive_senders = responsive_clients.entry(address).or_default();

            for sender in senders {
                if sender
                    .send(sse::Event::Comment("ping".into()))
                    .await
                    .is_ok()
                {
                    responsive_senders.push(sender);
                }
            }
        }
```

Reply:
thanks, that makes sense! We also need to make sure that the empty vector for stale clients will get wiped out so perhaps the following would work?
```
        for (address, senders) in clients {
            if senders.is_empty() {
                continue;
            }
            let responsive_senders = responsive_clients.entry(address).or_default();
            for sender in senders {
                if sender
                    .send(sse::Event::Comment("ping".into()))
                    .await
                    .is_ok()
                {
                    responsive_senders.push(sender);
                }
            }
        }
```

Reply:
Yes, additionally if no senders are responsive, then the entry shouldn't be inserted.